shared/src/ledger/tx.rs

```diff
@@ -1916,6 +1927,102 @@ pub async fn build_custom<
+    args: args::GenIbcShieldedTransafer,
+) -> Result<Option<IbcShieldedTransfer>> {
+    let key = match args.target.payment_address() {
+        Some(pa) if pa.is_pinned() => Some(pa.hash()),
+        Some(_) => None,
+        None => return Ok(None),
+    };
+    let source = Address::Foreign(args.sender.clone());
+    let (src_port_id, src_channel_id) =
+        get_ibc_src_port_channel(client, &args.port_id, &args.channel_id)
+            .await?;
+    let token = namada_core::ledger::ibc::received_ibc_token(
+        &args.token,
+        args.trace_path,
+        &src_port_id,
+        &src_channel_id,
+        &args.port_id,
+        &args.channel_id,
+    )
+    .map_err(|e| {
+        Error::Other(format!("Getting IBC Token failed: error {e}"))
+    })?;
+    let validated_amount = validate_amount(client, args.amount, &token, false)
+        .await
+        .expect("expected to validate amount");
+
+    let shielded = shielded
+        .gen_shielded_transfer(
+            client,
+            &TransferSource::Address(source.clone()),
+            &args.target,
+            &token,
+            validated_amount,
+        )
+        .await
+        .map_err(|err| TxError::MaspError(err.to_string()))?;
+    let transfer = token::Transfer {
+        source: source.clone(),
+        target: masp(),
+        token,
+        amount: validated_amount,
+        key,
+        shielded: None,
+    };
+    if let Some(shielded) = shielded {
+        Ok(Some(IbcShieldedTransfer {
+            transfer,
+            masp_tx: shielded.masp_tx,
+        }))
+    } else {
+        Ok(None)
+    }
+}
+
+async fn get_ibc_src_port_channel<C: crate::ledger::queries::Client + Sync>(
+    client: &C,
+    dest_port_id: &PortId,
+    dest_channel_id: &ChannelId,
+) -> Result<(PortId, ChannelId)> {
+    use crate::ibc::core::ics04_channel::channel::ChannelEnd;
+    use crate::ibc_proto::protobuf::Protobuf;
+
+    let channel_key = channel_key(dest_port_id, dest_channel_id);
+    match rpc::query_storage_value_bytes::<C>(client, &channel_key, None, false)
+        .await
+    {
+        Ok((Some(bytes), _)) => {
+            let channel = ChannelEnd::decode_vec(&bytes).map_err(|_| {
+                Error::Other(format!(
+                    "Decoding channel end failed: port {}, channel {}",
+                    dest_port_id, dest_channel_id
+                ))
+            })?;
+            if let Some(src_channel) = channel.remote.channel_id() {
+                Ok((channel.remote.port_id.clone(), src_channel.clone()))
+            } else {
+                Err(Error::Other(format!(
+                    "The source channel doesn't exist: port {dest_port_id}, \
+                     channel {dest_channel_id}"
+                )))
+            }
+        }
+        _ => Err(Error::Other(format!(
+            "Reading channel end failed: port {dest_port_id}, channel \
+             {dest_channel_id}"
+        ))),
+    }
+}
```

Comment:
doesn't read very well. combining `Result` and `Option` methods functions better for error handling IMO